### PR TITLE
Swift 3 support

### PR DESCRIPTION
**Pull request by Sega-Zero**
Full swift 3 support based on Cereal 2.0, finally :)
depends on #25

### CODE REVIEW

File: Cereal/CerealDecoder.swift [hunk 1]
```diff
@@ -42,14 +42,14 @@ public struct CerealDecoder {
     private let reverseRange: StrideThrough<Int>
     private let indexMap: [String: Int]?
 
-    private func itemForKey(key: String) -> CoderTreeValue? {
+    fileprivate func itemForKey(key: String) -> CoderTreeValue? {
```
Comment (ketzusaka) [inline Cereal/CerealDecoder.swift]: This should be `itemForKey(_ key: String) -> CoderTreeValue?`. We shouldn't need to duplicate the word `key`.

**Comment by ketzusaka** [comment]
I think the enums should of remained without an argument label for the associated values. I think the enum is descriptive enough, and the documentation that Apple gives doesn't have an argument label on associated values: https://developer.apple.com/library/prerelease/content/documentation/Swift/Conceptual/Swift_Programming_Language/Enumerations.html

File: Cereal/CerealDecoder.swift (hunk 1 above)
Comment (Sega-Zero) [reply Cereal/CerealDecoder.swift]: I think this should be `itemFor(key:)`, according to new swift guidelines, will patch it asap

Comment (ketzusaka) [reply Cereal/CerealDecoder.swift]: Hmm, I'm fine with either, but `itemForKey(_` is more in line with how Apple imports `NSCoder`

Comment (Sega-Zero) [reply Cereal/CerealDecoder.swift]: But it's a `fileprivate` method :) Also, `NSCoder` has no itemForKey method, afaik.

**Comment by Sega-Zero** [comment]
I thought I removed all the associated value labels, where did I missed it?

**Comment by ketzusaka** [comment]
Oh god, nevermind that. They are removed on the enums. I mistook `CerealDecoder.____` as an enum case. In any case, I think things like `CerealDecoder.rootCerealItemsWithData` should actually be `CerealDecoder.rootCerealItems(with data: Data)`

**Comment by Sega-Zero** [comment]
Agree. Renaming will take some extra time :)

**Comment by Sega-Zero** [comment]
Should I rename CerealEncoder's `dataWithRootItem(_ root: ItemType) throws -> Data {` to
`data(with rootItem: ItemType) throws -> Data {`?

**Comment by ketzusaka** [comment]
Yeah, and all the other `dataWith______` calls please :)

**Comment by Sega-Zero** [comment]
Had to rename to `data(withRoot:)`, because without `root` in parameter name it becomes unclear what kind of value the function expects.

File: Cereal/Cereal.swift [hunk 2]
```diff
@@ -55,13 +55,13 @@ Cereal.register(Fooable.self)
 
 - parameter identifyingCerealType:  The type being registered.
 */
-public func register(identifyingCerealType: IdentifyingCerealType.Type) {
+public func register(_ identifyingCerealType: IdentifyingCerealType.Type) {
     identifierMap[identifyingCerealType.initializationIdentifier] = identifyingCerealType
 }
 
-func identifyingCerealTypeWithIdentifier(identifier: String) throws -> IdentifyingCerealType.Type {
+func identifyingCerealTypeWithIdentifier(_ identifier: String) throws -> IdentifyingCerealType.Type {
```
Comment (eddiekaiger) [inline Cereal/Cereal.swift]: this should be `identifyingCerealType(withIdentifier identifier: String)`

File: WatchExample Extension/InterfaceController.swift [hunk 3]
```diff
@@ -16,57 +16,57 @@ class EmployeeRowType: NSObject {
...
-            if let row = employeesTable.rowControllerAtIndex(index) as? EmployeeRowType {
+        for (index, employee) in employees.enumerated() {
+            if let row = employeesTable.rowController(at: index) as? EmployeeRowType {
                 row.nameLabel.setText(employee.name)
             }
         }
     }
 
-    private func updateEmployeesWithData(employeeData: NSData) {
+    fileprivate func updateEmployeesWithData(_ employeeData: Data) {
```
Comment (eddiekaiger) [inline WatchExample Extension/InterfaceController.swift]: `updateEmployees(withData employeeData: Data)`

**Comment by Sega-Zero** [comment]
Is there anything else? :)

File: Cereal/CerealDecoder.swift [hunk 4]
```diff
@@ -42,14 +42,14 @@ public struct CerealDecoder {
     private let reverseRange: StrideThrough<Int>
     private let indexMap: [String: Int]?
 
-    private func itemForKey(key: String) -> CoderTreeValue? {
+    fileprivate func itemFor(key: String) -> CoderTreeValue? {
```
Comment (eddiekaiger) [inline Cereal/CerealDecoder.swift]: `item(forKey: String)`

File: Cereal/CerealDecoder.swift [hunk 5]
```diff
@@ -1075,59 +1075,59 @@ public struct CerealDecoder {
...
+            throw CerealError.invalidEncoding("Failed to decode value \(value)")
         }
 
         return decodedResult
     }
 
     /// Helper method to improve decoding speed for large objects:
     /// method returns index map to read value for key without enumerating through all the items
-    private static func propertiesIndexMapForType(type: Any.Type, value: CoderTreeValue) -> [String: Int] {
+    fileprivate static func propertiesIndexMapForType(type: Any.Type, value: CoderTreeValue) -> [String: Int] {
```
Comment (eddiekaiger) [inline Cereal/CerealDecoder.swift]: `propertiesIndexMap(for type: Any.Type, value: CoderTreeValue)`

File: Cereal/CerealDecoder.swift [hunk 6]
```diff
@@ -26,7 +26,7 @@ private class TypeIndexMapHolder: SyncHolder<TypeIndexMap> {
         return result
     }
 
-    func save(map map: [String: Int], forType type: Any.Type) {
+    func save(map: [String: Int], forType type: Any.Type) {
```
Comment (eddiekaiger) [inline Cereal/CerealDecoder.swift]: the label here can just be `for`

**Review by eddiekaiger — CHANGES_REQUESTED** [review]

**Comment by Sega-Zero** [comment]
Hope you guys merge this PR someday =)
If no more changes required, please release pod version for swift-2, swift-2.3 and the latest swift-3

**Comment by ketzusaka** [comment]
Will-do. I'll look again today -- eddie's on vacation :)

**Review by ketzusaka — APPROVED** [review]
Looks good! Will merge tomorrow and get tags setup.

**Pull request merged**